### PR TITLE
Metrics

### LINKED ISSUE #2: Implement a Metrics class to capture algorithm performance.
For a better understanding of algorithms, a metrics class needs to be implemented which will potentially spit these out to a .csv file for further study. This class will keep track of important events within simulations to aid in the future implementation of a sceneless run of algorithms.

Should include the following:
- [ ] `history` : an array of values where `history[t]` is the value of the metric at time t.
- [ ] `freq` : an integer saying "calculate this metric every freq activations"... also need something maybe per rounds? vs. per activations?
- [ ] A function `calculate()` that takes as input a `AmoebotSystem` that calculates whatever it was supposed to calculate. Pure virtual, meaning definition signature ends with `= 0;`
- [ ] Build metrics that all `AmoebotSystems` have, like #iterations, #rounds, #moves.
- [ ] Display all metrics with ListView.

### PR DESCRIPTION
**Pull request by ZiadAbdelkarim**
From Joseph's original PR:

Created a base Metric class. Implemented two types of Metrics: Counts and Measures. Counts are values that increase monotonically and are tracked every round (e.g., moves, rounds, etc.) while Measures are system properties that are calculated less frequently, determinted by a fixed frequency. All Metrics are shown in the GUI.

Basic Counts (# Rounds, # Moves, and # Activations) have all been implemented as default.

Data can be output to .csv by using the "Data" button or with `Cmd`/`Ctrl` + M.

Resolves #2.

### CODE REVIEW

**Comment by jdaymude** [comment]
Resolves issue #2.

**Comment by jdaymude** [comment]
Updated export to use the JSON format (instead of CSV) and is performed with `Ctrl`/`Cmd + E`.

**Comment by jdaymude** [comment]
@josephbriones @ryanyiu This PR is now ready for review. The main contributions are:
- Adding `Count` and `Measure` classes in `core/metrics.*`.
- Replacing the hardcoded default metrics (# rounds and # moves) with three new counts: # Rounds, # Activations, and # Moves.
- Adding `PerimeterMeasure` to `alg/compression.*` to demonstrate how custom measures are created and linked to particle systems.
- Polishing the front-end GUI (fixing previous issues with overlapping text).
- Switching the data export format to JSON, modifying a few of the hotkeys, and updating the Usage documentation accordingly.
- Adding a new `ScriptInterface` command for exporting metrics JSON.

You can split up primary/secondary reviewing responsibilities however you'd like. Please be aware that there's one small issue we'll need to fix in the future:
- In the JSON data, I have a field `"algorithm"` that is supposed to identify the algorithm that was being tracked. However, our architecture is blocking me from getting this information easily, since only `ui/algorithm` knows the actual algorithm names. So currently, I have the JSON saying `"algorithm" : "???"`.

File: core/amoebotsystem.h [hunk 1]
```diff
@@ -52,40 +53,44 @@ class AmoebotSystem : public System, public RandomNumberGenerator {
...
+  // at least once, this resets its logging and triggers registerRound(), which
+  // commits all counts and measures to their histories and increments the
+  // number of completed asynchronous rounds by one.
+  void registerMovement(unsigned int numMoves = 1);
   void registerActivation(AmoebotParticle* particle);
+  void registerRound();
+
+  // Various access functions for metrics (counts and measures). getCounts
+  // (resp., getMeasures) returns a reference to the count (resp., measure)
+  // list. getCount (resp., getMeasure) return a reference to the named count
```
Comment (ryanyiu) [inline core/amoebotsystem.h]: Small change: `returns` instead of `return`

File: core/simulator.h [hunk 2]
```diff
@@ -31,22 +32,33 @@ class Simulator : public QObject {
...
-  int numMovements() const;
-  int numRounds() const;
+  QVariant metrics() const;
+
+  // Responds to the exportMetrics signal from the GUI and scripts by creating
+  // an output file with a unique timestamp (to avoid accidental overwrites) and
+  // writing the metrics JSON to it.
+  void exportMetrics();
+
+  // TODO.
```
Comment (ryanyiu) [inline core/simulator.h]: Did you want to keep the comment? Or should a GitHub issue be made for implementing this part (so the comment + function can be moved to a separate PR)?

**Comment by ryanyiu** [comment]
@jdaymude Everything looks good apart from one or two minor things, which I left comments on.

File: core/amoebotsystem.h (hunk 1 above)
Comment (jdaymude) [reply core/amoebotsystem.h]: I've addressed this in a new commit.

File: core/simulator.h (hunk 2 above)
Comment (jdaymude) [reply core/simulator.h]: All the functions here were already implemented; it's just that `simulator.h` didn't have any header function comments. I've added one for `saveScreenshotSetup` in the most recent commit.

**Review by josephbriones — COMMENTED** [review]
It seems as though I am unable to generate the .json metrics file. Currently trying to see if this is a Mac issue.

Additionally, would you like all the copyrights updated to 2020?

**Review by ryanyiu — CHANGES_REQUESTED** [review]
Can't seem to remove this comment, but everything was good (I got `json` files to generate on my end).

**Review by ryanyiu — DISMISSED** [review]
LGTM

File: core/simulator.cpp [hunk 3]
```diff
@@ -78,12 +79,34 @@ int Simulator::numObjects() const {
...
   QMutexLocker locker(&system->mutex);
-  return system->numRounds();
+  QFile outFile(QDir::currentPath() + "/metrics_" +
+                QString::number(QDateTime::currentSecsSinceEpoch()) + ".json");
+  if (!outFile.open(QIODevice::WriteOnly | QIODevice::Text)) {
+    return;
+  }
+  QTextStream outStream(&outFile);
+  outStream << system->metricsAsJSON();
+}
```
Comment (jdaymude) [inline core/simulator.cpp]: @josephbriones No need to update the copyrights in this PR, though I should definitely do that soon. As for the JSON generation, not sure what would be going wrong. This is the relevant code; as you can see in lines 100-101, everything is being done in platform-independent Qt libraries. According to the documentation, `"/"` is always translated to the correct directory separator if a `QFile` is used.

Comment (jdaymude) [reply core/simulator.cpp]: @josephbriones What happens when you try to generate the file? Does it crash? Or does it just not appear where you expect it to?

Comment (jdaymude) [reply core/simulator.cpp]: After talking to Joseph, we found that a JSON file is indeed being written. However, `QDir::currentPath()` on macOS is pointing at `<builddir>/AmoebotSim.app/Contents/MacOS`, which is a hidden directory that isn't directly accessible by Finder. I'm investigating how we could patch this to point at a stable location on Windows and macOS so that users will know where to go look.

Comment (jdaymude) [reply core/simulator.cpp]: Ok, I think I've found a solution I'm happy with. @josephbriones try replacing `QDir::currentPath()` with `QCoreApplication::applicationDirPath()`. This should be the `QString` path for the directory containing the executable, which is stable and visible on both Windows and macOS. If this works for you, I'll go ahead and push a new commit.

**Review by ryanyiu — APPROVED** [review]
LGTM. As an aside, should we open up a separate GitHub issue for building some kind of notification to a user when a metric file is made? It feels a bit unresponsive to just press the button, and not see anything telling you that metric data has been created.

**Comment by jdaymude** [comment]
@ryanyiu actually, that's exactly what happens when you use the `exportMetrics()` command on the command line. But emitting this signal from the button requires a lot of signal/slot piping that I thought was unnecessary at this point in time. Yet another thing to improve in the UI after we figure out #7 and #3...

**Review by josephbriones — APPROVED** [review]
LGTM.
